test_node_env/expected_writer.js

```diff
@@ -0,0 +1,18 @@
+import { Script } from 'vm';
+import { createWriteStream } from 'fs';
+
+let writer2 = createWriteStream("test_node_env/expected.txt");
+const args = process.argv.slice(2);
+const expected = args[0];
+
+writer2.once('open', function(fd) {
+    let r = e.runInNewContext();
```

Comment:
Is there a reason variable `e` is declared after the writer is opened? Since the writer depends on variable `e`, it is possible that `e` is undefined when called here.

Reply:
The reason why variable `e` is declared after the writer is opened is that I just want to make changes at the last line. It still works fine so I don't think that e can be undefined.

Reply:
> The reason why variable `e` is declared after the writer is opened is that I just want to make changes at the last line. It still works fine so I don't think that `e` can be undefined.

Not true at all. It works on your computer, but this doesn't mean it will still work on a faster computer, or computers with a single core which are unable to open the writer and assign the variable in parallel. 

A simple example to proof this, we can simply add a delay before declaring the variable.
```diff
writer.once("open", function (fd) {
+  console.log("Function called", s);
  let r = s.runInThisContext();
  if (typeof r !== "undefined") {
    writer.write(JSON.stringify(r));
  } else {
    writer.write("undefined");
  }
  writer.end();
});

+ const delayTime = 10;
+ const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
+ await delay(delayTime).then(console.log(`${delayTime}ms delay over`, "\n"));

let s = new Script(readFileSync(programPath));
```

On my local machine, when `delayTime` is >= 5ms, s is undefined and throws an error.

What's actually happening here:
You are relying on the assumption that 
1) opening the writer and assigning the variable is done in parallel.
2) opening the writer takes longer than defining the variable.

However, this assumption does not always hold. Always best to declare variables before you use them.

Reply:
Oh I understand, would it be oke if I change the code to this?
<img width="501" alt="Screenshot 2021-07-30 at 11 32 31 AM" src="https://user-images.githubusercontent.com/77202230/127596004-c4cbabe9-5e76-4a00-8352-f2e398aef419.png">

Reply:
Yes this would work!

Reply:
Thank you, I fixed it.